[inductor] Use 64-bit indexing for large tensors in triton code

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97447
* #98608
* #98607

This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

should fix #96978, #93606 (need to double check)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[TESTING] Use 64-bit indexing in inductor triton code

[ghstack-poisoned]

peterbell10 committed:
Update on "Use 64-bit indexing for large tensors in inductor triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

[ghstack-poisoned]

weiwangmeta commented:
Sorry, we only occasionally allow two 5hr jobs running on 8 runners unless it is really necessary. See this job https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml 
Please help do a rebase (which disabled long-job trigger label) and trigger it only when necessary (like https://github.com/pytorch/pytorch/actions/runs/4536573015). 

Alternatively, there is a ciflow/inductor-perf-compare/ label that runs for ~1 hour if you want to check performance.

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1626,6 +1638,56 @@ def reduction_hint(node):
+        buf_numel = [sympy_product(buf.get_size()) for buf in buffers]
+        buf_storage_sizes = [buf.get_layout().storage_size() for buf in buffers]
+        for numel, storage_size in zip(buf_numel, buf_storage_sizes):
+            if V.graph.sizevars.size_hint(numel) > int_max:
+                return False
```
What's an example of an operation where the indexing operations overflow, and `numel > int_max` but not ` storage_size > int_max`?

peterbell10 replied:
The case I'm concerned about is broadcasting into a reduction, e.g.

```python
@torch.compile
def fn(a, b):
    return torch.sum(a + b)
```

where `a` has shape `(1, s0)` and `b` has shape `(s0, 1)`.

The generated reduction loop looks like:
```python
for roffset in range(0, rnumel, RBLOCK):
    rindex = roffset + rbase
    rmask = rindex < rnumel
    r1 = (rindex // ks0)
    r0 = rindex % ks0
    tmp0 = tl.load(in_ptr0 + (r1), rmask, eviction_policy='evict_last', other=0)
    tmp2 = tl.load(in_ptr1 + (r0), rmask, eviction_policy='evict_last', other=0)
    tmp2 = tmp0 + tmp1
    _tmp3 = tl.where(rmask, _tmp3 + tmp2, _tmp3)

```

You can see that `rindex` has a maximum value of `s0*s0` even though each input buffer only stores `s0` elements. Although I just tested it out and it looks like the input doesn't actually have the expanded shape... So it might be better to use the kernel's numel instead.

lezcano replied:
Ah, right, we don't create independent indices for these. Note that that happens in your example without the `torch.sum`  as well. Just doing `a.unsqueeze(1) + b` for vectors `a`, `b` gets you that issue. 
Now, when it comes to where to look at for the size, I think you're right, and looking at `xnumel, rnumel` may be the safer thing to do. The intuition here being that we know that we start reading garbage from the pointers when the masks are false, that is, when the computations go over `{xyz}mask` or `rmask`. Actually, perhaps via this reasoning, we just need to test that these are no larger than `int_max` and we should be good? Also, since we split reductions, perhaps we just need to check `xnumel`?

peterbell10 replied:
The reduction is necessary so the output buffer isn't larger than `int_max`, otherwise the `storage_size` check is enough. 

>  Actually, perhaps via this reasoning, we just need to test that these are no larger than `int_max` and we should be good?

If by "these" you mean `xnumel` and `rnumel` then that doesn't hold for strided indexing calculations.

>  Also, since we split reductions, perhaps we just need to check `xnumel`?

This only holds if the splitting installs guards, which I don't think is correct.

lezcano replied:
After an offline discussion, this looks fine. Let's leave a comment with a few points that were discussed, in particular why checking `xnumel * rnumel` and `storage_size` should be good enough.

peterbell10 committed:
Fix cpu on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

[ghstack-poisoned]

peterbell10 committed:
Fix ExprPrinter test on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -1479,6 +1480,95 @@ def fn(a, b):
+        self.assertTrue((actual == 4).all())
+
+    def test_large_strided_reduction(self):
+        # Test 64-bit indexing is used when input numel is less than INT_MAX
+        # but stride calculations go above INT_MAX
```
Why do strides computation go above `INT_MAX`? I thought we discussed that stride computations can just be as large as the size of the underlying tensor. This should be the case here as well.

peterbell10 replied:
The tensor storage is larger than `INT_MAX` here, but `numel < storage_size` because the tensor is non-dense.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1626,6 +1638,61 @@ def reduction_hint(node):
+        if any(size_hint(size) > int_max for size in buf_sizes):
+            return False
+
+        # Only install guards for 32-bit indexing as there is no correctness
+        # issue with using 64-bit for everything
+        for size in buf_sizes:
+            V.graph.sizevars.guard_leq(size, int_max)
```
Wouldn't this break if you first pass to `(x.unsqueeze(0)+y).sum()` two vectors of size `2**15` and then two vectors of size `2**16`, as it wouldn't recompile?

peterbell10 replied:
Agreed, I forgot to guard on numel

peterbell10 committed:
Fix off-by-one error and missing guard on numel on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@ngimel PTAL

ngimel commented:
Does this fix #96978 ?

ngimel commented on torch/_inductor/select_algorithm.py:
```diff
@@ -356,6 +369,14 @@ def generate(
 
+        numel = sympy_product(layout.size)
+        buffers = itertools.chain(input_nodes, (fake_out,))
+        index_dtype = (
```
for matmuls int64 indexing is usually pretty catastrophic in terms of perf (due to increased register pressure), so going forward it will be better to split matmuls to use int32 indexing always.

peterbell10 replied:
To be clear,  is this okay for now or would you prefer an assert?

ngimel replied:
Can you check perf of some big matrix? If it's bad I'd prefer an assert/fallback to aten, generating matmuls is purely perf optimization, we can just not do it for any reason.

peterbell10 replied:
The results were suspiciously similar so I looked at the triton code and of course the templates hard-code 32-bit indexing which in hindsight is obvious. I've changed it to not add the kernel as a choice to the autotuner when I detect 32-bit indexing would overflow.

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -523,7 +525,8 @@ def add(node):
         size = self.kernel.indexing_size_str(self.index, self.prefix)
-        return f"tl.arange(0, {self.prefix.upper()}BLOCK){size}"
+        index_dtype = self.kernel.index_dtype
+        return f"tl.arange(0, {self.prefix.upper()}BLOCK){size}.to({index_dtype})"
```
I've changed `arange` and `program_id` to only add the `to` when necessary,

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -915,18 +922,22 @@ def indexing(
             if copy_shape:
-                index_str = f"{index_str} + tl.zeros({copy_shape}.shape, tl.int32)"
+                index_str = (
+                    f"{index_str} + tl.zeros({copy_shape}.shape, {self.index_dtype})"
```
I'm a bit concerned about `ops.index_expr` using 32-bit values in calculations that overflow, but I suppose a lot of lowerings currently hard-code 32-bit dtypes so are probably broken anyway.

lezcano replied:
Working in the indirect indexing stuff, I've seen that plenty of lowerings hardcode that. It'd be nice at some point to reduce the number of lowerings so that these fixes affect more functions directly, without having to modify the lowerings.

ngimel replied:
triton's `index_expr` overload ignores dtype anyway, for cpp it's a problem

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1693,6 +1696,9 @@ def __eq__(self, other) -> bool:
         )
 
+    def storage_size(self) -> sympy.Expr:
+        return sympy_dot(self.size, self.stride) + self.offset
```
changed to `_prims_common.compute_required_storage_length` which does the full calculation and guards against zeros.

peterbell10 commented:
> Does this fix https://github.com/pytorch/pytorch/issues/96978 ?

Running the forward kernel gives an illegal memory access for me still.

ngimel commented:
Hm that's weird, can you dig into that?

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
The IMA I was seeing was actually an issue with the benchmark wrapper which went away after a rebase (one of the pointers was null). I can't say for sure that #96978 is fixed since I get an OOM before it gets to the backwards, but the forwards kernel is working at least.

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

should fix #96978, #93606 (need to double check)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ngimel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1136,6 +1144,7 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
 
             if accumulator_index:
                 # argmax, argmin
+                idx_dtype = self.index_dtype
```
is it used anywhere?

peterbell10 committed:
Update on "[inductor] Use 64-bit indexing for large tensors in triton code"


This changes `TritonKernel` to have an `index_dtype` property which is
used as the dtype in indexing calculations. By default it is
`tl.int32` but if any input or output buffer is larger than `INT_MAX`
then we use `tl.int64` instead.

should fix #96978, #93606 (need to double check)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
